2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+# Naming: Climbing Towards Abstraction
+
+Your code and its artifacts are strewn with names. You name repositories, files, packages, classes, functions, variables, and on and on. The names you choose determine the quality of the conversations you have with your programmer peers.
+
+On the flip side of that bargain, your understanding of libraries and frameworks, i.e. other people's code, starts with names. Names form the substrate of APIs, delineating their boundaries, clarifying their inputs and outputs, and defining how they're used.
+
+If you can trust that a library works, you need look no further than its API docs to understand it. Ideally, API docs are the only point of contact you have with a third-party library. And they are, essentially, a bunch of names.
+
+Just as a writer grinds through word choices, a programmer must struggle with names. As the joke goes, the two hardest things in computer science are naming, cache invalidation, and off-by-one errors. No matter the programming language, no matter whether you're operating from a functional or object-oriented perspective, no matter if the language is statically or dynamically typed, names are a ubiquitous and critical part of your code. Cache invalidation may not matter in your application, but names sure do. The quality of your work depends on the quality of your names.
+
+When writing code, you spend a lot of time burrowed in the minutiae of specifics. You ply your craft at the mercy of unbending syntax rules &mdash; rules that will bring a million-line codebase to its knees over a single missing character. You write small functions that, given an exact input, produce an exact output. Every time.
+
+But naming demands that you wrest your mind from granular, concrete details and think more generally, more categorically.
+
+The linguist Sam Hayakawa, in his book [Language in Thought and Action](https://www.amazon.com/Language-Thought-Action-S-I-Hayakawa/dp/0156482401/), describes a model of language he calls the Abstraction Ladder. On this conceptual ladder, the bottommost rung holds the most concrete idea. The topmost rung is occupied by a decidedly abstract, intangible concept. In between, ideas get more abstract the higher up they are. The Abstraction Ladder is straightforward to understand, but studying it leads to broad and profound insights.
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
```

Comment:
This is a great diagram but I can't figure out what the bottom-most image is: a pair of lungs? Mittens and a zipper? Just wondering.

Reply:
Ha, now I'm wondering! I think it's lungs and a spine. At least that's the story I'm telling! I clarified in the caption in this commit: 1ec077264e27e2c778857aa09c3c3ac32d128b08.

Reply:
What I *thought* might get someone's pushback is the idea of a rat as a pet. But I'm going to double down on that!